File: riistaandroid/src/main/java/fi/riista/mobile/activity/MainActivity.java
package fi.riista.mobile.activity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ActivityInfo;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.location.LocationManager;
import android.net.ConnectivityManager;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.PopupMenu;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Window;

import com.roughike.bottombar.BottomBar;
import com.roughike.bottombar.OnTabReselectListener;
import com.roughike.bottombar.OnTabSelectListener;

import fi.riista.mobile.AppConfig;
import fi.riista.mobile.NetworkConnectivityReceiver;
import fi.riista.mobile.R;
import fi.riista.mobile.database.DiarySync.DiarySyncContext;
import fi.riista.mobile.database.GameDatabase;
import fi.riista.mobile.database.PermitManager;
import fi.riista.mobile.database.SpeciesInformation;
import fi.riista.mobile.pages.AnnouncementsFragment;
import fi.riista.mobile.pages.ContactDetailsFragment;
import fi.riista.mobile.pages.GameLogFragment;
import fi.riista.mobile.pages.HomeViewFragment;
import fi.riista.mobile.pages.MyDetailsFragment;
import fi.riista.mobile.pages.SettingsFragment;
import fi.riista.mobile.utils.AppPreferences;
import fi.riista.mobile.utils.Utils;

public class MainActivity extends BaseActivity implements DiarySyncContext {

    private GameDatabase mDatabase;

    private BroadcastReceiver mConnectivityReceiver = null;

    private BottomBar mBottomNavigationView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        requestWindowFeature(Window.FEATURE_INDETERMINATE_PROGRESS);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        super.onCreate(savedInstanceState);
        getSupportActionBar().setHomeAsUpIndicator(new ColorDrawable(Color.TRANSPARENT));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        setContentView(R.layout.activity_main);

        initDatabase();

        replacePageFragment(HomeViewFragment.newInstance());
        setupNavigationView();

        LocationManager manager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        if (!manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            createGPSPrompt();
        }

        SpeciesInformation.refreshInfo(this);

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(ConnectivityManager.CONNECTIVITY_ACTION);
        mConnectivityReceiver = new NetworkConnectivityReceiver(getWorkContext());
        registerReceiver(mConnectivityReceiver, intentFilter);

        if (mDatabase.getSyncMode(this) == GameDatabase.SyncMode.SYNC_AUTOMATIC) {
            mDatabase.initSync(getWorkContext(), 500);
        }

        if (getIntent().getBooleanExtra(LoginActivity.SHOW_ANNOUNCEMENTS_EXTRA, false)) {
            getIntent().removeExtra(LoginActivity.SHOW_ANNOUNCEMENTS_EXTRA);
            mBottomNavigationView.selectTabWithId(R.id.menu_announcements);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        mDatabase.stopSyncing();

        if (mConnectivityReceiver != null) {
            unregisterReceiver(mConnectivityReceiver);
        }

        mDatabase.close();
    }

    @Override
    protected void onResume() {
        super.onResume();

        // Handle resuming sync when returning to app
        if (mDatabase.getSyncMode(this) == GameDatabase.SyncMode.SYNC_AUTOMATIC) {
            mDatabase.initSync(getWorkContext(), 500);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);

        //savedInstanceState.putInt("menuitem", mSelectedItem.ordinal());
    }

    public void initDatabase() {
        mDatabase = GameDatabase.getInstance();
        mDatabase.load(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        int selectedId = mBottomNavigationView.getCurrentTabId();

        if (selectedId == R.id.menu_main) {
            finish();
        } else {
            mBottomNavigationView.selectTabWithId(R.id.menu_main);
        }
    }

    private void setupNavigationView() {
        mBottomNavigationView = (BottomBar) findViewById(R.id.bottom_navigation);
        mBottomNavigationView.setOnTabSelectListener(new OnTabSelectListener() {
            @Override
            public void onTabSelected(@IdRes int tabId) {
                Fragment fragment = null;

                switch (tabId) {
                    case R.id.menu_main:
                        fragment = HomeViewFragment.newInstance();
                        break;
                    case R.id.menu_game_log:
                        fragment = GameLogFragment.newInstance();
                        break;
                    case R.id.menu_announcements:
                        fragment = AnnouncementsFragment.newInstance();
                        break;
                    case R.id.menu_my_details:
                        fragment = MyDetailsFragment.newInstance();
                        break;
                    case R.id.menu_more:
                        displayMorePopupMenu();
                        break;
                }

                replacePageFragment(fragment);
            }
        });
        mBottomNavigationView.setOnTabReselectListener(new OnTabReselectListener() {
            @Override
            public void onTabReSelected(@IdRes int tabId) {
                switch (tabId) {
                    case R.id.menu_more:
                        displayMorePopupMenu();
                        break;
                }
            }
        });
    }

    private void displayMorePopupMenu() {
        PopupMenu popupMenu = new PopupMenu(MainActivity.this, mBottomNavigationView.getTabWithId(R.id.menu_more));
        popupMenu.getMenuInflater().inflate(R.menu.main_navigation_more, popupMenu.getMenu());
        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                Fragment innerFragment = null;

                switch (item.getItemId()) {
                    case R.id.menu_contact:
                        innerFragment = ContactDetailsFragment.newInstance();
                        break;
                    case R.id.menu_settings:
                        innerFragment = SettingsFragment.newInstance();
                        break;
                    case R.id.menu_magazine:
                        Intent intent = new Intent(MainActivity.this, MagazineActivity.class);
                        if (AppPreferences.LANGUAGE_CODE_SV.equalsIgnoreCase(AppPreferences.getLanguageCodeSetting(MainActivity.this))) {
                            intent.putExtra(MagazineActivity.EXTRA_URL, AppConfig.MAGAZINE_URL_SV);
                        }
                        else {
                            intent.putExtra(MagazineActivity.EXTRA_URL, AppConfig.MAGAZINE_URL_FI);
                        }
                        startActivity(intent);
                        break;
                    case R.id.menu_logout:
                        confirmLogout();
                        break;
                }

                replacePageFragment(innerFragment);
                return true;
            }
        });

        popupMenu.setGravity(Gravity.RIGHT);
        popupMenu.show();
    }

    void replacePageFragment(Fragment fragment) {
        if (fragment != null) {
            FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
            fragmentTransaction.replace(R.id.content_frame, fragment);
            fragmentTransaction.commit();
        } else {
            Utils.LogMessage("Fragment null");
        }
    }

    private void selectItem(int position) {
        mBottomNavigationView.selectTabAtPosition(position);
    }

    public void confirmLogout() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(getString(R.string.logout) + "?")
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        logout();
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        // User canceled. Do nothing
                    }
                });
        builder.show();
    }

    public void logout() {
        mDatabase.stopSyncing();
        mDatabase.clearUpdateTimes();
        mDatabase.removeCredentials(this);
        AppPreferences.setUserInfo(this, null);
        PermitManager.getInstance(this).clearPermits();
        Utils.unregisterNotificationsAsync();

        Intent intent = new Intent(MainActivity.this, LoginActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }

    public void createGPSPrompt() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(getResources().getString(R.string.gps_prompt));
        builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int id) {
                startActivity(new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS));
            }
        });
        builder.setNegativeButton(R.string.no, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int id) {
                // Do nothing
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    @Override
    public void syncStarted() {
        setSyncNotification(true);
    }

    @Override
    public void syncCompleted() {
        setSyncNotification(false);
    }

    /**
     * Use to recreate activity to apply updated language setting
     */
    public void restartActivity() {
        Intent intent = getIntent();
        mDatabase.close();
        finish();
        startActivity(intent);
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/activity/MainActivity.kt
package fi.riista.mobile.activity

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.location.LocationManager
import android.net.ConnectivityManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import com.google.android.material.bottomnavigation.BottomNavigationView
import dagger.android.AndroidInjection
import fi.riista.common.RiistaSDK
import fi.riista.common.domain.userInfo.LoginStatus
import fi.riista.common.reactive.AppObservable
import fi.riista.common.reactive.DisposeBag
import fi.riista.common.reactive.disposeBy
import fi.riista.mobile.ExternalUrls.Companion.getEventSearchUrl
import fi.riista.mobile.ExternalUrls.Companion.getHuntingSeasonsUrl
import fi.riista.mobile.NetworkConnectivityReceiver
import fi.riista.mobile.R
import fi.riista.mobile.database.HarvestDatabase
import fi.riista.mobile.database.HarvestDatabaseMigrationToRiistaSDK
import fi.riista.mobile.database.SpeciesInformation
import fi.riista.mobile.feature.groupHunting.GroupHuntingActivity
import fi.riista.mobile.feature.huntingControl.HuntingControlActivity
import fi.riista.mobile.feature.login.LoginActivity
import fi.riista.mobile.feature.moreView.MoreItemType
import fi.riista.mobile.feature.moreView.MoreViewFragment
import fi.riista.mobile.feature.unregister.UnregisterUserAccountActivityLauncher
import fi.riista.mobile.observation.ObservationDatabase
import fi.riista.mobile.observation.ObservationDatabaseMigrationToRiistaSDK
import fi.riista.mobile.pages.AnnouncementsFragment
import fi.riista.mobile.pages.ContactDetailsFragment
import fi.riista.mobile.pages.GalleryFragment
import fi.riista.mobile.pages.GameLogFragment
import fi.riista.mobile.pages.HomeViewFragment
import fi.riista.mobile.pages.MapViewer
import fi.riista.mobile.pages.MapViewer.FullScreenExpand
import fi.riista.mobile.pages.MyDetailsFragment
import fi.riista.mobile.pages.SettingsFragment
import fi.riista.mobile.pages.ShootingTestCalendarEventListFragment
import fi.riista.mobile.srva.SrvaDatabaseMigrationToRiistaSDK
import fi.riista.mobile.sync.AnnouncementSync
import fi.riista.mobile.sync.AppSync
import fi.riista.mobile.sync.AppSync.AppSyncListener
import fi.riista.mobile.ui.AlertDialogFragment
import fi.riista.mobile.ui.DelegatingAlertDialogListener
import fi.riista.mobile.ui.AlertDialogId
import fi.riista.mobile.utils.AppPreferences
import fi.riista.mobile.utils.BackgroundOperationStatus
import fi.riista.mobile.utils.CredentialsStore
import fi.riista.mobile.utils.LogoutHelper
import fi.riista.mobile.utils.UserInfoStore
import fi.riista.mobile.utils.Utils
import fi.riista.mobile.utils.openInBrowser
import fi.riista.mobile.utils.toVisibility
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import javax.inject.Inject

class MainActivity
  : BaseActivity(),
    AppSyncListener,
    FullScreenExpand,
    MoreViewFragment.InteractionManager
{
    @Inject
    lateinit var harvestDatabase: HarvestDatabase

    @Inject
    lateinit var observationDatabase: ObservationDatabase

    @Inject
    lateinit var appSync: AppSync

    @Inject
    lateinit var logoutHelper: LogoutHelper

    @Inject
    lateinit var announcementsSync: AnnouncementSync

    @Inject
    lateinit var credentialsStore: CredentialsStore

    @Inject
    lateinit var userInfoStore: UserInfoStore

    @Inject
    lateinit var backgroundOperationStatus: BackgroundOperationStatus

    @Suppress("unused")
    private val dialogListener = DelegatingAlertDialogListener(this).apply {
        registerPositiveCallback(AlertDialogId.MAIN_ACTIVITY_LOGOUT_CONFIRMATION) {
            logout()
        }
        registerPositiveCallback(AlertDialogId.MAIN_ACTIVITY_GPS_PROMPT) {
            startActivity(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS))
        }
    }

    private var connectivityReceiver: BroadcastReceiver? = null
    private lateinit var bottomNavigationView: BottomNavigationView
    private lateinit var progressBar: ProgressBar
    private var syncOnResume = true
    override var groupHuntingAvailable = AppObservable(false)
    override var huntingControlAvailable = AppObservable(false)
    private val disposeBag = DisposeBag()

    public override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)
        supportActionBar?.setHomeAsUpIndicator(ColorDrawable(Color.TRANSPARENT))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        setContentView(R.layout.activity_main)
        progressBar = findViewById(R.id.progress_horizontal)
        checkIntentForSyncExtra(intent)
        initDatabase()
        replacePageFragment(HomeViewFragment())
        setupNavigationView()
        val manager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
        if (!manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
            createGPSPrompt()
        }
        SpeciesInformation.refreshInfo(this)

        // checkIntentForSyncExtra() must be called before registerConnectivityReceiver() because
        // syncOnResume is set in the former method.
        registerConnectivityReceiver()
        checkIntentForAnnouncementExtra(intent)
        registerLoginStatus()
        registerGroupHuntingStatus()
        registerHuntingControlStatus()
        indicateBackgroundOperationStatus()
    }

    private fun registerLoginStatus() {
        RiistaSDK.currentUserContext.loginStatus.bindAndNotify { loginStatus ->
            if (loginStatus is LoginStatus.LoggedIn) {
                checkGroupHuntingAvailability()

                // ensure hunting control gets updated data from the backend by refreshing
                checkHuntingControlAvailability(refresh = true)

                UnregisterUserAccountActivityLauncher.launchIfAccountUnregistrationRequested(parentActivity = this)
            }
        }.disposeBy(disposeBag)
    }

    private fun registerGroupHuntingStatus() {
        RiistaSDK.currentUserContext.groupHuntingContext
            .clubContextsProvider.loadStatus.bindAndNotify {
                groupHuntingAvailable.set(RiistaSDK.currentUserContext.groupHuntingContext.groupHuntingAvailable)
            }
            .disposeBy(disposeBag)
    }

    private fun checkGroupHuntingAvailability() {
        CoroutineScope(Dispatchers.Main).launch {
            RiistaSDK.currentUserContext.groupHuntingContext.checkAvailabilityAndFetchClubs()
        }
    }

    private fun registerHuntingControlStatus() {
        RiistaSDK.huntingControlContext
            .huntingControlRhyProvider.loadStatus.bindAndNotify {
                huntingControlAvailable.set(RiistaSDK.huntingControlContext.huntingControlAvailable)
            }
            .disposeBy(disposeBag)
    }

    private fun indicateBackgroundOperationStatus() {
        backgroundOperationStatus.backgroundOperationInProgress
            .bindAndNotify { backgroundOperationInProgress ->
                progressBar.visibility = backgroundOperationInProgress.toVisibility()
            }
            .disposeBy(disposeBag)
    }

    private fun checkHuntingControlAvailability(refresh: Boolean = false) {
        CoroutineScope(Dispatchers.Main).launch {
            RiistaSDK.huntingControlContext.checkAvailability(refresh)
        }
    }

    private fun checkIntentForSyncExtra(intent: Intent) {
        val doNotInitiateSync = intent.getBooleanExtra(DO_NOT_INITIATE_SYNC, false)
        syncOnResume = !doNotInitiateSync
    }

    private fun initDatabase() {
        val credentials = credentialsStore.get()
        val username = credentials?.username
        harvestDatabase.setUser(username)

        // Run migrations from app databases to Riista SDK database
        backgroundOperationStatus.startOperation(BackgroundOperationStatus.Operation.DATABASE_MIGRATION)
        SrvaDatabaseMigrationToRiistaSDK.copyEvents {
            ObservationDatabaseMigrationToRiistaSDK.copyObservations(observationDatabase) {
                HarvestDatabaseMigrationToRiistaSDK.copyHarvests(harvestDatabase) {
                    backgroundOperationStatus.finishOperation(BackgroundOperationStatus.Operation.DATABASE_MIGRATION)
                    appSync.enableSyncPrecondition(AppSync.SyncPrecondition.DATABASE_MIGRATION_FINISHED)
                }
            }
        }
    }

    fun replacePageFragment(fragment: Fragment?) {
        if (fragment != null) {
            val fragmentTransaction = supportFragmentManager.beginTransaction()
            fragmentTransaction.replace(R.id.content_frame, fragment)
            fragmentTransaction.addToBackStack(null)
            fragmentTransaction.commit()
        } else {
            Utils.LogMessage("Fragment null")
        }
    }

    private fun setupNavigationView() {
        bottomNavigationView = findViewById(R.id.bottom_navigation)
        bottomNavigationView.setOnNavigationItemSelectedListener { menuItem: MenuItem ->
            val fragment = when (menuItem.itemId) {
                R.id.menu_home -> HomeViewFragment()
                R.id.menu_game_log -> GameLogFragment()
                R.id.menu_map -> createMapFragment()
                R.id.menu_announcements -> AnnouncementsFragment.newInstance(true)
                R.id.menu_more -> createMoreViewFragment()
                else -> null
            }
            replacePageFragment(fragment)
            true
        }
    }

    private fun createMoreViewFragment(): Fragment {
        checkGroupHuntingAvailability()
        checkHuntingControlAvailability()
        return MoreViewFragment()
    }

    private fun createMapFragment(): Fragment {
        val fragment = MapViewer.newInstance(enablePois = true)
        // Map fragment is used in many places with different titles, so set the title manually.
        setCustomTitle(getString(R.string.map_title))
        return fragment
    }

    override fun moreItemClicked(type: MoreItemType) {
        when (type) {
            MoreItemType.MY_DETAILS -> replacePageFragment(MyDetailsFragment.newInstance())
            MoreItemType.GALLERY -> replacePageFragment(GalleryFragment.newInstance())
            MoreItemType.CONTACT_DETAILS -> replacePageFragment(ContactDetailsFragment.newInstance())
            MoreItemType.SHOOTING_TESTS -> replacePageFragment(ShootingTestCalendarEventListFragment())
            MoreItemType.SETTINGS -> replacePageFragment(SettingsFragment.newInstance())
            MoreItemType.HUNTING_DIRECTOR -> startActivity(Intent(this, GroupHuntingActivity::class.java))
            MoreItemType.HUNTING_CONTROL -> startActivity(Intent(this, HuntingControlActivity::class.java))
            MoreItemType.EVENT_SEARCH -> Uri.parse(getEventSearchUrl(languageCode)).openInBrowser(this)
            MoreItemType.MAGAZINE -> startActivity(MagazineActivity.getLaunchIntent(this))
            MoreItemType.SEASONS -> Uri.parse(getHuntingSeasonsUrl(languageCode)).openInBrowser(this)
            MoreItemType.LOGOUT -> confirmLogout()
        }
    }

    private val languageCode: String
        get() = AppPreferences.getLanguageCodeSetting(this)

    fun selectItem(id: Int) {
        bottomNavigationView.menu.findItem(id).isChecked = true
    }

    private fun confirmLogout() {
        AlertDialogFragment.Builder(this, AlertDialogId.MAIN_ACTIVITY_LOGOUT_CONFIRMATION)
            .setMessage(getString(R.string.logout) + "?")
            .setPositiveButton(android.R.string.ok)
            .setNegativeButton(android.R.string.cancel)
            .build()
            .show(supportFragmentManager)
    }

    private fun logout() {
        logoutHelper.logout(context = this)

        val intent = Intent(this, LoginActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(intent)
    }

    private fun createGPSPrompt() {
        AlertDialogFragment.Builder(this, AlertDialogId.MAIN_ACTIVITY_GPS_PROMPT)
            .setMessage(getString(R.string.gps_prompt))
            .setPositiveButton(R.string.yes)
            .setNegativeButton(R.string.no)
            .build()
            .show(supportFragmentManager)
    }

    private fun registerConnectivityReceiver() {
        if (syncOnResume) {
            val intentFilter = IntentFilter()
            intentFilter.addAction(ConnectivityManager.CONNECTIVITY_ACTION)
            Log.d("MainActivity", "Registering connectivity receiver")
            connectivityReceiver = NetworkConnectivityReceiver(appSync)
            registerReceiver(connectivityReceiver, intentFilter)
        }
    }

    private fun checkIntentForAnnouncementExtra(intent: Intent) {
        if (intent.getBooleanExtra(LoginActivity.SHOW_ANNOUNCEMENTS_EXTRA, false)) {
            intent.removeExtra(LoginActivity.SHOW_ANNOUNCEMENTS_EXTRA)
            announcementsSync.sync {

                // A state check is needed because user might switch to another Activity before
                // sync task finishes.
                if (lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) {
                    replacePageFragment(AnnouncementsFragment.newInstance(false))
                }
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        checkIntentForSyncExtra(intent)
        registerConnectivityReceiver()
        checkIntentForAnnouncementExtra(intent)
    }

    override fun onStart() {
        super.onStart()
        appSync.addSyncListener(listener = this, notifyImmediately = true)
    }

    override fun onResume() {
        super.onResume()

        // Handle resuming sync when returning to app.
        if (syncOnResume) {
            appSync.enableSyncPrecondition(AppSync.SyncPrecondition.HOME_SCREEN_REACHED)
        }

        UnregisterUserAccountActivityLauncher.launchIfAccountUnregistrationRequested(parentActivity = this)
    }

    override fun onStop() {
        super.onStop()
        appSync.removeSyncListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()

        appSync.disableSyncPrecondition(AppSync.SyncPrecondition.HOME_SCREEN_REACHED)

        if (connectivityReceiver != null) {
            unregisterReceiver(connectivityReceiver)
        }
        harvestDatabase.close()

        disposeBag.disposeAll()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        val selectedId = bottomNavigationView.selectedItemId
        if (selectedId == R.id.menu_home) {
            // task may live longer, ensure notification is displayed again when resuming
            UnregisterUserAccountActivityLauncher.resetCooldown()
            finish()
        } else {
            bottomNavigationView.selectedItemId = R.id.menu_home
        }
    }

    override fun setFullscreenMode(fullscreen: Boolean) {
        val attrs = window.attributes
        val actionBar = supportActionBar
        if (fullscreen) {
            attrs.flags = attrs.flags or WindowManager.LayoutParams.FLAG_FULLSCREEN
            bottomNavigationView.visibility = View.GONE
            actionBar?.hide()
        } else {
            attrs.flags = attrs.flags and WindowManager.LayoutParams.FLAG_FULLSCREEN.inv()
            bottomNavigationView.visibility = View.VISIBLE
            actionBar?.show()
        }
        window.attributes = attrs
    }

    override fun onSyncStarted() {
        backgroundOperationStatus.startOperation(BackgroundOperationStatus.Operation.SYNCHRONIZATION)
    }

    override fun onSyncCompleted() {
        backgroundOperationStatus.finishOperation(BackgroundOperationStatus.Operation.SYNCHRONIZATION)
    }

    /**
     * Use to recreate activity to apply updated language setting.
     */
    fun restartActivity() {
        val intent = intent
        harvestDatabase.close()
        finish()
        startActivity(intent)
    }

    companion object {
        const val DO_NOT_INITIATE_SYNC = "doNotInitiateSync"
    }
}
